engine/resources/hetzner.go

```diff
@@ -0,0 +1,716 @@
+// instances. A resource in the `absent` state only exists as a local
+// blueprint, and does not exist as server instance on Hetzner's side.
+// Note: for now, all manually registered SSH keys are used when creating a
+// server. More selective use to be added soon.
+type HetznerRes struct {
+	traits.Base
+	init *engine.Init
+
+	// APITokenPath is the path of a file containing the project's API token.
+	// Provide the full path, incl. filetype, and keep the token private.
+	// Consider where the hcloud CLI stores tokens in default operation.
+	APITokenPath string `lang:"apitokenpath"`
+	// ServerType determines how much ram, disk and cpu the resource has.
+	Servertype string `lang:"servertype"`
+	// Datacenter determines where the resource is hosted.
+	Datacenter string `lang:"datacenter"`
+	// Image determines the image the resource is booted with.
+	Image string `lang:"image"`
+
+	// State is the desired steady state of the server instance.
+	// Choose between "running", "off" or "absent".
+	State string `lang:"state"`
+
+	// WaitInterval is the sleep interval in seconds that is used when waiting
+	// for transient states to converge between intermediate operations.
+	// Note that the hetzner API supports bursts at up to 3600 requests per hour.
+	// Nonzero values disable request bursts, a zero value will allow it.
+	// A wait interval >= 1s is recommended for extended use.
+	WaitInterval uint32 `lang:"waitinterval"`
+
+	// ServerRescueMode specifies the image used when enabling rescue mode.
+	// The supported images are "linux32", "linux64" and "freebsd64".
+	// Alternatively, use "disabled" to disable rescue mode (default).
+	ServerRescueMode string `lang:"serverrescuemode"`
+	// ServerRescueSSHKeys is used to select a subset of keys that
+	// should be available for rescue mode operations over ssh.
+	// From all ssh keys known to the project client, choose a subset of
+	// keys by name, as an array of strings. New keys need to be added
+	// manually via the cloud console web api for now.
+	// If a keyname is unknown to the client, it is ignored.
+	ServerRescueSSHKeys []string `lang:"serverrescuekeys"`
+
+	// Full client details for interaction with the hcloud-go API.
+	client *hcloud.Client
+	// Snapshot of the server instance, for checkapply operations.
+	server *hcloud.Server
+	// Snapshot of the serverCreateOpts struct generated with hcloud.
+	serverconfig hcloud.ServerCreateOpts
+	// hcloud-formatted array of ssh keys for use with rescue mode.
+	rescuekeys []*hcloud.SSHKey
+
+	// Internal flag to simplify rescue mode operations in checkapply.
+	// False if ServerRescueMode="disabled", true otherwise.
+	rescueEnable bool
+}
+
+func (obj *HetznerRes) Default() engine.Res {
+	return &HetznerRes{
+		WaitInterval:     WaitIntervalLimit,
+		ServerRescueMode: "disabled",
+	}
+}
+
+func (obj *HetznerRes) Validate() error {
+
+	// verify name (TODO: how to ensure unique names across server instances?)
```

Comment:
What do you mean about this TODO?

Reply:
I was not sure if I need to add explicit protection against name collisions (eg. two servers of the same name specified in a .mcl file under the same API token). As I understand it now, mgmt seems to do a good job already in catching such name collisions when generating the resource graph. Probably okay as is then?

Reply:
Correct-- mgmt will not allow to identical kind+name pairs. The hetzner resource will never see another hetzner resource with the exact same name. But it could see a different kind of resource with the same Name().